x-pack/plugins/enterprise_search/public/applications/app_search/components/engine/engine_router.tsx

```diff
@@ -68,12 +68,19 @@ export const EngineRouter: React.FC = () => {
 
   const { engineName: engineNameFromUrl } = useParams() as { engineName: string };
   const { engineName, dataLoading, engineNotFound } = useValues(EngineLogic);
-  const { setEngineName, initializeEngine, clearEngine } = useActions(EngineLogic);
+  const { setEngineName, initializeEngine, pollEmptyEngine, stopPolling, clearEngine } = useActions(
+    EngineLogic
+  );
 
   useEffect(() => {
     setEngineName(engineNameFromUrl);
     initializeEngine();
-    return clearEngine;
+    pollEmptyEngine();
```

Comment:
Did you consider making `pollEmptyEngine` a side effect of calling `initializeEngine`?

Currently, the UI needs to call `initializeEngine`, call `pollEmptyEngine`, and also call `stopPolling`. It'd be great if the UI could just call `initializeEngine` and the polling automatically started and stopped when appropriate.

initializeEngine:
  - Fetch Engine data
  - If the Engine is empty, and not already polling, start polling
  - If the engine is not empty, stop polling

You could use an `unmount` hook in Kea to ensure that polling is cancelled whenever the logic unmounts: https://kea.js.org/docs/guide/additional/#events

**Or better yet**

Maybe we don't even need to consider the polling state...

initializeEngine:
  - Fetch Engine data
  - If the Engine is empty, wait x second and call initializeEngine again

Just thinking out loud.

Reply:
I always personally waffle between making things more explicit and easier for devs to follow in code, or making things "magical" and less code but potentially harder to follow when reading/debugging.

> You could use an unmount hook in Kea to ensure that polling is cancelled whenever the logic unmounts: https://kea.js.org/docs/guide/additional/#events

FYI, we can't do this. I tried this initially, and what's happening is EngineLogic is **not** fully unmounting between engine changes (primarily meta engine->source engine navigation), it's only clearing state, so the unmount hook does not get fired. You do need to very explicitly call a specific action in `useEffect`. If you skip this call you end with a million intervals not properly getting cleared.

> if the Engine is empty, wait x second and call initializeEngine again

Going to strongly push back against using timeouts over setting an interval. We've done this in the past and it has previously lead to [duplicate API calls](https://github.com/elastic/kibana/pull/95732#discussion_r603717946) - we had a discussion in Slack a while back about it, and most of the frontend devs who replied were in favor of moving away from `setTimeout` for polling over `setInterval`. In additional, it's a significantly clearer from an API signal standpoint that a poll is happening when you use `setInterval` over `setTimeout`.

Reply:
Ah, yeah. We ran into that issue before. Marius's recommendation was to [key the logic and use BindLogic](https://github.com/keajs/kea/issues/120#issuecomment-778658935) so that it gets reset when the engine changes.

Anyway, this still looks like it could be simplified to me, but you're in this deeper than me though, I'm not familiar with all of the guts and history, so if you think this is the best path then I'm fine with it.

Reply:
Ahh that would definitely solve that issue & I'd super be down to key EngineLogic sometime! Probably going to punt on it for 7.14 FF for the sake of time if that's cool (currently focusing on UI/UX polish over strictly dev/tech debt only items). I've made a backlog item though: https://github.com/elastic/app-search-team/issues/1853

I think doing so actually lets us refactor out or "is engine stale" check which is super cool. Definitely wanna do this for 7.15 😎

Reply:
Totally! Also, wanted to say I really like the way you implemented it. Putting it in the EngineLogic is dope, we can take advantage of that in like every view without actually doing anything. Super elegant.